### PR TITLE
Fix being able to set zero or negative hitobject duration via timeline

### LINKED ISSUE #10959: Sliders can have zero length
Sliders and spinners can be modified to have zero length. Sliders by using transformation tools, and spinners by dragging in the timeline. It looks like spinners can also be dragged into negative values (as seen on the video where only the number is left)

**video:** https://youtu.be/ypFaliLLgi8
**osu!lazer version:** 2020.1121.0

### PR DESCRIPTION
**Pull request by vmaggioli**
Moving control points on sliders allowed for the slider to have no length, and timeline bars for spinners could be dragged to their start time in the timeline for zero length as well.

**test video** (same test as in issue): https://youtu.be/u_q7MUUEeAg

Fixes part of #10959

### CODE REVIEW

File: osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs:390]: I am not sure why the added comparison is written this way. I assume it's because there's a worry of a precision issue, but I don't know whether it is actually an issue here, if `IBeatSnapProvider` quantises properly. In other words, if comparing directly was buggy, the preceding comparison would also be.

It would be good to have test coverage for this as well, if you can figure that out.

**Review by bdach — DISMISSED** [review]

File: osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs
Comment (vmaggioli) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs:390]: I can work on getting a test class set up for this. For the comparison, I was noticing a precision issue when I was debugging. I was noticing that the if the `hitObject.StartTime` had a value of something like 16825 then `snappedTime` often times would be something like 16825.12756..., so I just added the subtraction comparison to ensure it still matches

Comment (bdach) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs:390]: In that case, instead of explicitly doing that math, this should use `Precision.AlmostEquals` from framework (grep around for other usages). For both comparisons even, I'd say. Otherwise it looks as if these checks are half-buggy.

Comment (vmaggioli) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs]: Changed to use `Precision.AlmostEquals`. I'll start the test class work later today

Comment (vmaggioli) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs]: Resolving this conversation as the `AlmostEquals` has been added

**Comment by vmaggioli** [comment]
Currently needing a bit of help with the testing, I've got the class set up properly I believe but I'm not finding any way with the current layout to get the location of the drag bar to move the mouse to. Any suggestions?
https://github.com/vmaggioli/osu/blob/e4b59c7317b7788f03b184ccc900aee717153658/osu.Game.Tests/Visual/Editing/TestSceneTimelineHitObjectBlueprint.cs

File: osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs
Comment (bdach) [reply osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineHitObjectBlueprint.cs]: no need to add comments about it. would have eventually been resolved at some point anyway, most likely when I (or anyone else) is re-reviewing after tests have been added.

**Comment by bdach** [comment]
The `.ChildrenOfType<T>()` extension method should be sufficient to locate the drag bar in the hierarchy, at which point you can access its `ScreenSpaceDrawQuad` if needed. There are many usages of such a pattern in other tests.

**Comment by vmaggioli** [comment]
thanks! that comment helped me get the test set up. last issue I haven't been able to figure out is how to get the `load` method for the `TimelineTestScene` to run. Debugging shows that the method never gets invoked and I see the `BackgroundDependencyLoader` attribute from the [osu-framework](https://github.com/ppy/osu-framework/blob/master/osu.Framework/Allocation/BackgroundDependencyLoaderAttribute.cs), but I'm assuming I'm missing some call to get the view drawn to invoke the `load` method...

**Comment by bdach** [comment]
I'm not sure what you mean by that. On checking out this branch in its current state (which I'm guessing doesn't have your latest changes yet), breakpoints on `TimelineTestScene.load()` work as expected. Could you explain the issue you're having or push code so that I can debug it without having to guess?

**Comment by vmaggioli** [comment]
For sure, just pushed the latest changes. My concern is the `load` method here
https://github.com/ppy/osu/blob/82848a7d70e8bb1dbd7bed2b4a178dfe2ce94bcc/osu.Game.Tests/Visual/Editing/TimelineTestScene.cs#L27

**Comment by bdach** [comment]
I've looked at your code, and a few things:

* The `load()` method runs here. Maybe it's not running when you're expecting it to; a test scene's `load()` runs before any test, as part of the test's initialisation.
* Maybe it was too much to expect a full test outright - the setup you were going for didn't really have much chance of working (was adding the timeline blueprint directly to the timeline, while it's intended to be added through a `TimelineBlueprintContainer`). I've adjusted the test so that it works properly now.
* I've also looked at your other test (for the zero-length-slider case), and it was also broken, due to adding test steps inside of test steps, which doesn't work - the tests are added at the end, so the test wasn't actually moving any path control points. In fact, when I adjusted the test to work properly, it turned out that your solution wasn't working, so it will *purposefully* fail now.

Please look through the attached changes and see if you can figure out how to fix the zero-length slider case properly.

**Comment by vmaggioli** [comment]
What I've found so far is that the `ExpectedDistance` of the `SliderPath` is still 300 up until `invalidate()` is called and the `version` is incremented. After that, the next line I can check a reference to the HitObject to is pasted below, and by that point the `Distance` is already 0. Still not 100% sure where the update occurs. It's definitely happening before the code changes I made to prevent an update though.
https://github.com/vmaggioli/osu/blob/d0fd2ae432700dc8a7d125b827e6931ab69940c2/osu.Game.Rulesets.Osu/Skinning/Default/SnakingSliderBody.cs#L95

**Comment by bdach** [comment]
@vmaggioli have you looked into this at all since your last comment? if not, are you going to?

**Comment by bdach** [comment]
Because the linked issue is milestoned for January I had a look at this again. I think any reasonably working solution has to do something like the following patch:

```diff
diff --git a/osu.Game/Rulesets/Objects/SliderPath.cs b/osu.Game/Rulesets/Objects/SliderPath.cs
index 3083fcfccb..8c2dcef85d 100644
--- a/osu.Game/Rulesets/Objects/SliderPath.cs
+++ b/osu.Game/Rulesets/Objects/SliderPath.cs
@@ -167,9 +167,28 @@ private void ensureValid()
             if (pathCache.IsValid)
                 return;
 
+            tryApplyNewPath();
+        }
+
+        private void tryApplyNewPath()
+        {
+            var previousPath = calculatedPath.ToList();
+            var previousCumulativeLength = cumulativeLength.ToList();
+            var previousLength = calculatedLength;
+
             calculatePath();
             calculateLength();
 
+            if (calculatedLength < 1)
+            {
+                // rollback changes.
+                calculatedPath.Clear();
+                calculatedPath.AddRange(previousPath);
+                cumulativeLength.Clear();
+                cumulativeLength.AddRange(previousCumulativeLength);
+                calculatedLength = previousLength;
+            }
+
             pathCache.Validate();
         }
```

That is not nearly enough, since the knowledge that the path update was refused should be propagated back up to the editor's components (like the path control point pieces themselves), and also the 1ms limit is honestly sort of arbitrary. That goes for the `IHasDuration` thing in here too - there's not much of a practical difference between a 1ms spinner/slider and a 0ms one. So overall:

* I'd be in favour of tearing out the slider part of this PR, as it seems too large in scope to handle cleanly here
* While the `IHasDuration` fix prevents negative durations, the lower limit should probably be rethought - maybe the minimum should be one beat in beat snap terms?

**Comment by peppy** [comment]
I'd definitely be for having this PR only address the timeline change, and looking at the slider blueprint on a fresh slate. Definitely needs some thought.

**Comment by vmaggioli** [comment]
@bdach sorry, I've been dealing with some issues with work over the past week so I haven't been attentive to my changes. I'll work on tearing out the slider path changes and keeping the timeline changes tonight. Apologies for the delay

**Comment by vmaggioli** [comment]
I've removed the slider changes so they can be dealt with in a separate PR/issue

**Comment by bdach** [comment]
I'm still not sure about the 1ms figure choice, but other than that this seems sane enough...

Dismissing my review, would appreciate a second pair of eyes on this.

**Comment by frenzibyte** [comment]
Yeah setting the minimum duration to one beat at least sounds better than an arbitrary 1ms.

**Comment by vmaggioli** [comment]
I believe I've switched it from 1 ms to 1 beat properly, but let me know if it's incorrect/not the optimal way

**Review by bdach — APPROVED** [review]
looks fine to me

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**
